### PR TITLE
[WIP; RFC, discussion, feedback] account for ashift when choosing buffers …

### PR DESCRIPTION
**Pull request by kernelOfTruth**
…to be written to l2arc device

If we don't account for that, then we might end up overwriting disk
area of buffers that have not been evicted yet, because l2arc_evict
operates in terms of disk addresses.

The discrepancy between the write size calculation and the actual increment
to l2ad_hand was introduced in
commit e14bb3258d05c1b1077e2db7cf77088924e56919

Also, consistently use asize / a_sz for the allocated size, psize / p_sz
for the physical size.  Where the latter accounts for possible size
reduction because of compression, whereas the former accounts for possible
size expansion because of alignment requirements.

The code still assumes that either underlying storage subsystems or
hardware is able to do read-modify-write when an L2ARC buffer size is
not a multiple of a disk's block size.  This is true for 4KB sector disks
that provide 512B sector emulation, but may not be true in general.
In other words, we currently do not have any code to make sure that
an L2ARC buffer, whether compressed or not, which is used for physical I/O
has a suitable size.

References:
https://reviews.csiden.org/r/112/diff/1/#index_header
https://github.com/zfsonlinux/zfs/pull/3433

Ported-by:
kernelOfTruth kernelOfTruth@gmail.com

Fixes (supposedly):
zfsonlinux#3114
zfsonlinux#3400
[and potentially several others]

### CODE REVIEW

**Comment by kernelOfTruth** [comment]
@avg-I your updated patch reaped some great results:

https://github.com/zfsonlinux/zfs/issues/3400#issuecomment-109654781

https://github.com/zfsonlinux/zfs/issues/3400#issuecomment-110467679

**Comment by pzwahlen** [comment]
@kernelOfTruth Would it be possible to update this patch against current master ? I'm willing to run that stack for a few days...

[mockbuild@dev-c7 gitzfs]$ patch -p1 < 3451.patch  
patching file module/zfs/arc.c
Hunk #1 succeeded at 5776 with fuzz 1 (offset 755 lines).
Hunk #2 FAILED at 5050.
Hunk #3 FAILED at 5095.
Hunk #4 succeeded at 5865 with fuzz 1 (offset 755 lines).
Hunk #5 FAILED at 5132.
Hunk #6 succeeded at 5930 (offset 756 lines).
Hunk #7 succeeded at 5941 (offset 756 lines).
Hunk #8 succeeded at 5958 (offset 756 lines).
3 out of 8 hunks FAILED -- saving rejects to file module/zfs/arc.c.rej

Thanks again for that great work...

**Comment by avg-I** [comment]
BTW, I've updated the patch against the latest illumos code: https://reviews.csiden.org/r/229/

**Comment by avg-I** [comment]
Ah, I've missed #3491 - thanks!

**Comment by kernelOfTruth** [comment]
@pzwahlen the updated fix is at https://github.com/zfsonlinux/zfs/pull/3491

@avg-I seems like I need to update the commit message again =)

Thanks you !

**Comment by kernelOfTruth** [comment]
closing, superseded by https://github.com/zfsonlinux/zfs/pull/3521
